### ISSUE TITLE
Target project does not exist: npm:rimraf@2.4.5

### ISSUE THREAD

**Issue by markabrahams** [nrwl/nx#20421]
### Current Behavior

When I run:
`npx nx run myproject:serve`

I get the error message:
```
>  NX   An error occured while creating pruned lockfile

   Please open an issue at `https://github.com/nrwl/nx/issues/new?template=1-bug.yml` and provide a reproduction.
   To prevent the build from breaking we are returning the root lock file.
   If you run `npm install --package-lock-only` in your output folder it will regenerate the correct pruned lockfile.
   
Original error: Target project does not exist: npm:rimraf@2.4.5


   Error: Target project does not exist: npm:rimraf@2.4.5
    at validateCommonDependencyRules (/home/mark/myproject/node_modules/nx/src/project-graph/project-graph-builder.js:323:15)
    at validateDependency (/home/mark/myproject/node_modules/nx/src/project-graph/project-graph-builder.js:313:5)
    at ProjectGraphBuilder.addDependency (/home/mark/myproject/node_modules/nx/src/project-graph/project-graph-builder.js:190:9)
    at ProjectGraphBuilder.addStaticDependency (/home/mark/myproject/node_modules/nx/src/project-graph/project-graph-builder.js:89:14)
    at /home/mark/myproject/node_modules/nx/src/plugins/js/lock-file/project-graph-pruning.js:81:17
    at Array.forEach (<anonymous>)
    at traverseNode (/home/mark/myproject/node_modules/nx/src/plugins/js/lock-file/project-graph-pruning.js:78:36)
    at /home/mark/myproject/node_modules/nx/src/plugins/js/lock-file/project-graph-pruning.js:80:9
    at Array.forEach (<anonymous>)
    at traverseNode (/home/mark/myproject/node_modules/nx/src/plugins/js/lock-file/project-graph-pruning.js:78:36)

```

So I'm opening this as per the above request in the error message.

Note that this doesn't occur on the initial compile, but does occur on every recompile after that.

Also note that this doesn't *seem* to have any impact on the actual serving of the app, which seems fine.

### Expected Behavior

No reported error message.

### Steps to Reproduce

1. Run command:

`npx nx run myproject:serve`

2. Edit a file and save.  On recompile the error is produced.

### Nx Report

```shell
mark@hope:~/myproject$ npx nx report
>  NX  Falling back to ts-node for local typescript execution. This may be a little slower.
  - To fix this, ensure @swc-node/register and @swc/core have been installed

 >  NX   Report complete - copy this into the issue template

   Node   : 18.17.1
   OS     : linux-x64
   npm    : 9.6.7
   
   nx                 : 17.1.3
   @nx/js             : 17.1.3
   @nx/jest           : 17.1.3
   @nx/linter         : 17.1.3
   @nx/eslint         : 17.1.3
   @nx/workspace      : 17.1.3
   @nx/devkit         : 17.1.3
   @nx/esbuild        : 17.1.3
   @nx/eslint-plugin  : 17.1.3
   @nx/node           : 17.1.3
   @nrwl/tao          : 17.1.3
   @nx/webpack        : 17.1.3
   typescript         : 5.2.2
```

### Operating System

- [ ] macOS
- [X] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

I haven't yet attempted a repo reproduction due to lack of time at this point, but can try this at some point.

Note that this repo has gone through several upgrades to Nx, so it could be that some artefacts from previous versions causing problems?

Here's the project.json file for the "myproject" app:

```
{
    "name": "myproject",
    "$schema": "../../node_modules/nx/schemas/project-schema.json",
    "sourceRoot": "apps/myproject/src",
    "projectType": "application",
    "targets": {
        "build": {
            "executor": "@nx/webpack:webpack",
            "outputs": ["{options.outputPath}"],
            "options": {
                "target": "node",
                "compiler": "tsc",
                "outputPath": "dist/apps/myproject",
                "main": "apps/myproject/src/main.ts",
                "tsConfig": "apps/myproject/tsconfig.app.json",
                "assets": ["apps/myproject/src/assets"],
                "generatePackageJson": true
            },
            "configurations": {
                "production": {
                    "optimization": true,
                    "extractLicenses": true,
                    "inspect": false,
                    "fileReplacements": [
                        {
                            "replace": "apps/myproject/src/environments/environment.ts",
                            "with": "apps/myproject/src/environments/environment.prod.ts"
                        }
                    ]
                }
            }
        },
        "serve": {
            "executor": "@nx/js:node",
            "options": {
                "buildTarget": "myproject:build",
                "port": 9239
            },
            "configurations": {
                "production": {
                    "buildTarget": "myproject:build:production"
                }
            }
        },
        "lint": {
            "executor": "@nx/eslint:lint",
            "outputs": ["{options.outputFile}"],
            "options": {
                "lintFilePatterns": ["apps/myproject/**/*.ts"]
            }
        },
        "test": {
            "executor": "@nx/jest:jest",
            "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
            "options": {
                "jestConfig": "apps/myproject/jest.config.ts"
            }
        }
    },
    "tags": []
}

```

**Comment by vonangelo** [comment]
I'm experiencing same type of error after migrating from 15.9.4 to 17.x.x. Removing "generatePackageJson" from project.json -> targets -> build solves this for me, although it's not a valid workaround, as I need to generate the package.json file.

**Comment by decline** [comment]
I also faced the same issue in a NestJS app after updating to Nx 17.1.3. Got the error every time the app recompiles due to file changes. @vonangelo's hint about the `generatePackageJson` field helped me to resolve the issue:

I just moved the `generatePackageJson` setting out of the default `build` target config in my `project.json`, and just moved it to the `production` configuration of the `build` target (as I don't need to generate the `package.json` in development mode).

So now serving works without an error and also building (with "production" configuration) also works without an error.

**Comment by vonangelo** [comment]
@decline Your change fixed it for me aswell! Although I'm curious to why it's necessary, as creating a new basic NX project with e.g. a React application and generatePackageJson setting enabled and located in the default build target works just fine. We didn't need to generate package.json on development, so it made sense to move it.

Thank you!

**Comment by KeithGillette** [comment]
In our project, the pruned lockfile complained about `Error: Target project does not exist: npm:isarray@1.0.0`, but does no longer since following @decline's suggestion of moving `generatePackageJson: true` out of the default `build` configuration in `project.json`. 

Now if `@nx/js:node` would reload the application properly after each successful webpack compilation instead of ever _other_ compilation following a file change ...

**Comment by JamesHenry** [comment]
Hi Folks, there have been a number of improvements in this area since the latest report.

Please can I request that someone tries with the latest Nx (19.2.2 at the time of writing) and reports back?

If it is still an issue, please provide minimal steps to reproduce.

Many thanks!

**Comment by valtersantosmatosmanulife** [comment]
@JamesHenry , something happen on 19.2.3, my case:

```
 NX   An error occured while creating pruned lockfile
Please open an issue at `https://github.com/nrwl/nx/issues/new?template=1-bug.yml` and provide a reproduction.
To prevent the build from breaking we are returning the root lock file.
If you run `npm install --package-lock-only` in your output folder it will regenerate the correct pruned lockfile.
Original error: Target project does not exist: npm:eventemitter3@3.1.2
Error: Target project does not exist: npm:eventemitter3@3.1.2
    at validateCommonDependencyRules (/Users/santova/Projects/gwam-asia-monorepo/node_modules/nx/src/project-graph/project-graph-builder.js:323:15)
    at validateDependency (/Users/santova/Projects/gwam-asia-monorepo/node_modules/nx/src/project-graph/project-graph-builder.js:313:5)
    at ProjectGraphBuilder.addDependency (/Users/santova/Projects/gwam-asia-monorepo/node_modules/nx/src/project-graph/project-graph-builder.js:190:9)
    at ProjectGraphBuilder.addStaticDependency (/Users/santova/Projects/gwam-asia-monorepo/node_modules/nx/src/project-graph/project-graph-builder.js:89:14)
    at /Users/santova/Projects/gwam-asia-monorepo/node_modules/nx/src/plugins/js/lock-file/project-graph-pruning.js:81:17
    at Array.forEach (<anonymous>)
    at traverseNode (/Users/santova/Projects/gwam-asia-monorepo/node_modules/nx/src/plugins/js/lock-file/project-graph-pruning.js:78:36)
    at /Users/santova/Projects/gwam-asia-monorepo/node_modules/nx/src/plugins/js/lock-file/project-graph-pruning.js:80:9
    at Array.forEach (<anonymous>)
    at traverseNode (/Users/santova/Projects/gwam-asia-monorepo/node_modules/nx/src/plugins/js/lock-file/project-graph-pruning.js:78:36)
```

    with:
```
 "dependencies": {
    "@apollo/client": "3.9.9",
    "@apollo/server": "4.10.2",
    "@azure/identity": "4.0.1",
    "@azure/msal-browser": "3.11.1",
    "@azure/service-bus": "7.9.4",
    "@azure/storage-blob": "12.17.0",
    "@nestjs/apollo": "12.1.0",
    "@nestjs/axios": "3.0.2",
    "@nestjs/common": "10.3.7",
    "@nestjs/core": "10.3.7",
    "@nestjs/graphql": "12.1.1",
    "@nestjs/platform-express": "10.3.7",
    "@nestjs/schedule": "4.0.1",
    "@nestjs/serve-static": "4.0.1",
    "@nestjs/swagger": "7.3.0",
    "@nestjs/typeorm": "10.0.2",
    "@salesforce-ux/design-system": "2.23.2",
    "@tanstack/match-sorter-utils": "8.11.8",
    "@tanstack/react-table": "8.15.0",
    "@types/postman-collection": "3.5.10",
    "adm-zip": "0.5.10",
    "ahooks": "3.7.10",
    "ajv": "8.12.0",
    "animejs": "3.2.2",
    "apollo-upload-client": "17.0.0",
    "await-to-js": "3.0.0",
    "axios": "1.6.8",
    "bignumber.js": "9.1.2",
    "body-parser": "1.20.2",
    "buffer": "6.0.3",
    "change-case": "5.4.3",
    "class-transformer": "0.5.1",
    "class-validator": "0.14.0",
    "classnames": "2.5.1",
    "colors": "1.4.0",
    "core-js": "3.36.1",
    "crypto-hash": "2.0.1",
    "csv": "6.3.8",
    "dataloader": "2.2.2",
    "deep-object-diff": "1.1.9",
    "diff": "5.2.0",
    "diff2html": "3.4.47",
    "exponential-backoff": "3.1.1",
    "flat": "5.0.2",
    "formik": "2.4.5",
    "framer-motion": "10.16.4",
    "gql-query-builder": "3.8.0",
    "graphiql": "3.2.1",
    "graphql": "16.8.1",
    "graphql-upload-minimal": "1.6.0",
    "handlebars": "4.7.8",
    "helmet": "7.1.0",
    "jose": "4.15.5",
    "jsonwebtoken": "9.0.2",
    "lodash": "4.17.21",
    "luxon": "3.4.4",
    "mime": "3.0.0",
    "mongodb": "5.9.2",
    "mssql": "10.0.2",
    "neverthrow": "6.1.0",
    "newrelic": "11.14.0",
    "nodemailer": "6.9.13",
    "postman-collection": "4.4.0",
    "prism-react-renderer": "2.3.1",
    "puppeteer": "22.6.1",
    "react": "18.3.1",
    "react-collapsible": "2.10.0",
    "react-dnd": "16.0.1",
    "react-dnd-html5-backend": "16.0.1",
    "react-dom": "18.3.1",
    "react-dropzone": "14.2.3",
    "react-loading-skeleton": "3.4.0",
    "react-number-format": "5.3.4",
    "react-query": "3.39.3",
    "react-router-dom": "6.22.3",
    "react-toastify": "9.1.3",
    "react-virtualized-auto-sizer": "1.0.24",
    "react-window": "1.8.10",
    "reflect-metadata": "0.1.13",
    "regenerator-runtime": "0.13.11",
    "rxjs": "7.8.1",
    "serialize-error": "8.1.0",
    "source-map-support": "0.5.21",
    "ssh2": "1.15.0",
    "ssh2-sftp-client": "9.1.0",
    "stacktrace-js": "2.0.2",
    "systeminformation": "5.22.6",
    "tmp": "0.2.3",
    "ts-class-initializable": "1.0.2",
    "typeorm": "0.3.20",
    "uuid": "9.0.1",
    "validator": "13.11.0",
    "winston": "3.13.0",
    "workerpool": "8.0.0",
    "ws": "8.16.0",
    "xirr": "1.1.0",
    "xlsx": "file:vendor/xlsx-0.20.2-patch.tgz",
    "yaml": "2.4.2",
    "yup": "1.4.0"
  },
  "devDependencies": {
    "@babel/core": "7.24.3",
    "@babel/preset-react": "7.24.1",
    "@graphql-codegen/cli": "4.0.1",
    "@graphql-codegen/typescript-operations": "4.2.0",
    "@graphql-codegen/typescript-react-apollo": "3.3.7",
    "@kubernetes-models/istio": "2.3.0",
    "@nestjs/schematics": "10.1.1",
    "@nestjs/testing": "10.3.7",
    "@nx/eslint": "19.2.3",
    "@nx/eslint-plugin": "19.2.3",
    "@nx/jest": "19.2.3",
    "@nx/js": "19.2.3",
    "@nx/nest": "19.2.3",
    "@nx/node": "19.2.3",
    "@nx/playwright": "19.2.3",
    "@nx/react": "19.2.3",
    "@nx/web": "19.2.3",
    "@nx/webpack": "19.2.3",
    "@nx/workspace": "19.2.3",
    "@playwright/test": "1.44.1",
    "@pmmmwh/react-refresh-webpack-plugin": "0.5.10",
    "@svgr/webpack": "8.1.0",
    "@testing-library/react": "15.0.6",
    "@types/adm-zip": "0.5.5",
    "@types/animejs": "3.1.12",
    "@types/apollo-upload-client": "17.0.5",
    "@types/body-parser": "1.19.5",
    "@types/diff": "5.0.9",
    "@types/express": "4.17.21",
    "@types/flat": "5.0.5",
    "@types/jest": "29.5.12",
    "@types/jsonwebtoken": "9.0.6",
    "@types/lodash": "4.17.0",
    "@types/luxon": "3.3.3",
    "@types/mime": "3.0.4",
    "@types/mssql": "8.1.2",
    "@types/multer": "1.4.11",
    "@types/newrelic": "9.14.3",
    "@types/node": "18.17.9",
    "@types/nodemailer": "6.4.14",
    "@types/opossum": "6.2.3",
    "@types/oracledb": "5.3.0",
    "@types/react": "18.3.1",
    "@types/react-dom": "18.3.0",
    "@types/react-onclickoutside": "6.7.10",
    "@types/react-outside-click-handler": "1.3.3",
    "@types/react-router-dom": "5.3.3",
    "@types/react-window": "1.8.8",
    "@types/ssh2-sftp-client": "9.0.3",
    "@types/supertest": "2.0.15",
    "@types/uuid": "9.0.8",
    "@typescript-eslint/eslint-plugin": "7.4.0",
    "@typescript-eslint/parser": "7.4.0",
    "autoprefixer": "10.4.19",
    "axios-mock-adapter": "1.22.0",
    "babel-jest": "29.7.0",
    "cross-env": "7.0.3",
    "csp-html-webpack-plugin": "5.1.0",
    "eslint": "8.57.0",
    "eslint-config-prettier": "9.1.0",
    "eslint-import-resolver-typescript": "3.6.1",
    "eslint-plugin-import": "2.29.1",
    "eslint-plugin-jsdoc": "48.2.2",
    "eslint-plugin-jsx-a11y": "6.8.0",
    "eslint-plugin-neverthrow": "1.1.4",
    "eslint-plugin-playwright": "0.18.0",
    "eslint-plugin-react": "7.34.1",
    "eslint-plugin-react-hooks": "4.6.0",
    "eslint-plugin-sonarjs": "0.23.0",
    "eslint-plugin-unicorn": "51.0.1",
    "eslint-plugin-unused-imports": "3.1.0",
    "glob": "9.3.0",
    "html-webpack-plugin": "5.6.0",
    "jest": "29.7.0",
    "jest-cucumber": "4.4.0",
    "jest-environment-jsdom": "29.7.0",
    "jest-environment-node": "29.7.0",
    "jest-html-reporter": "3.10.2",
    "kubernetes-models": "4.3.1",
    "mongodb-memory-server": "8.13.0",
    "nx": "19.2.3",
    "nx-stylelint": "17.1.4",
    "postcss": "8.4.38",
    "prettier": "3.2.5",
    "prettier-plugin-gherkin": "2.2.1",
    "prettier-plugin-sql": "0.16.0",
    "pretty-quick": "3.1.3",
    "react-refresh": "0.14.0",
    "rimraf": "4.4.0",
    "source-map-loader": "4.0.1",
    "sql.js": "1.10.2",
    "stylelint": "15.11.0",
    "stylelint-config-standard": "34.0.0",
    "supertest": "6.3.4",
    "ts-jest": "29.1.2",
    "ts-loader": "9.5.1",
    "ts-node": "10.9.2",
    "tsconfig-paths": "4.2.0",
    "tslib": "2.6.2",
    "typescript": "5.4.5",
    "url-loader": "4.1.1"
  },
  "optionalDependencies": {
    "@nx/nx-darwin-arm64": "17.0.3",
    "@nx/nx-darwin-x64": "17.0.3",
    "@nx/nx-linux-x64-gnu": "17.0.3",
    "@nx/nx-win32-x64-msvc": "17.0.3"
  },
  "overrides": {
    "react": "$react",
    "react-dom": "$react-dom",
    "@nx/webpack": {
      "@adobe/css-tools": "4.3.2"
    }
  }
```

**Comment by leosvelperez** [comment]
@valtersantosmatosmanulife, could you retry using the latest version? We released some more fixes in 19.3.0 that might address what you're facing.

If the issue persists, we will need a proper reproduction to troubleshoot it. These issues are tricky to reproduce, and it is hard to guess what the problem could be.
